processing/src/main/java/org/apache/druid/query/DefaultQueryMetrics.java

```diff
@@ -289,12 +320,26 @@ public void identity(String identity)
   }
 
   @Override
-  public void emit(ServiceEmitter emitter)
+  public void emit(final ServiceEmitter emitter)
```

Comment:
`emit()` still shouldn't be concurrent. It should allow emitting only from the owner thread.

Reply:
What's the benefits of maintaining QueryMetrics's owner thread? I think one benefit of making QueryMetrics thread-safe is that users can create it in one thread and use it in another thread without explicitly transferring ownership.

Reply:
This is the corner stone of why I originally made QueryMetrics single-threaded. I agree that the ability to add dimensions and metrics from parallel threads is convenient, but allowing to `emit()` from concurrent threads is paving the road to hell of literally undiscoverable and undebuggable bugs with metrics.

All exceptions related to QueryMetrics that have been floating up since the introduction of this class represent our own misunderstanding of the execution flow of some queries - how and when data is transferred between threads, in what specific thread pool some code is actually executed, etc. QueryMetrics force maintainers to refine the threading and make data transitions more explicit, that is a good thing. Making QueryMetrics permissive is easier but it's like burying head in the sand.

Where do you need parallel `emit()` and parallel `report()` is not enough?

Reply:
After checking the code again, I found that emit() is always called from a single thread, which could be considered the owner thread for the QueryMetrics instance. Thanks for your explanation!